### PR TITLE
Add ability to refetch a query in useQuery and useQueryWithStore

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: packages/ra-core/src/dataProvider/useQuery.ts [hunk 1]
```diff
@@ -113,19 +123,21 @@ const useQuery = (
                     : [payload, otherOptions]
             )
             .then(({ data, total }) => {
-                setState({
+                setState(prev => ({
+                    ...prev,
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useQuery.ts:127]: I remember that we used to do that but we had problems in corner cases. I suggest you pass `refetch` explicitly  instead.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Looking forward to reading the documention!

File: packages/ra-core/src/dataProvider/useQueryWithStore.ts [hunk 2]
```diff
@@ -114,19 +115,26 @@ const useQueryWithStore = <State extends ReduxState = ReduxState>(
...
     totalSelector: (state: State) => number = defaultTotalSelector(query),
     isDataLoaded: (data: any) => boolean = defaultIsDataLoaded
-): {
-    data?: any;
-    total?: number;
-    error?: any;
-    loading: boolean;
-    loaded: boolean;
-} => {
+): StateResult => {
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useQueryWithStore.ts:118]: rename to UseQueryWithStoreValue

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/dataProvider/useQueryWithStore.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/dataProvider/useQueryWithStore.ts:118]: It would be a breaking change, right?

Comment (fzaninotto) [reply packages/ra-core/src/dataProvider/useQueryWithStore.ts:118]: Why? The type wasn't named or exported before this PR.

File: packages/ra-core/src/dataProvider/index.ts [hunk 3]
```diff
@@ -25,12 +24,9 @@ import useDeleteMany from './useDeleteMany';
 import useRefreshWhenVisible from './useRefreshWhenVisible';
 import useIsAutomaticRefreshEnabled from './useIsAutomaticRefreshEnabled';
 
-export type {
-    QueryOptions,
-    UseMutationValue,
-    UseQueryValue,
-    UseGetOneHookValue,
-};
+export * from './useQueryWithStore';
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/index.ts:27]: This won't work: the `Query` type will conflict with the `Query` component.

The `Query` type name is too generic; Something like `DataProviderQuery` would be better, WDYT?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/dataProvider/useQuery.ts [hunk 4]
```diff
@@ -138,12 +151,6 @@ const useQuery = (
     return state;
 };
 
-export interface Query {
-    type: string;
-    resource?: string;
-    payload: object;
-}
-
 export interface QueryOptions {
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useQuery.ts:154]: This one should be named `UseQueryOptions`. Sorry to find the problems one by one, but the fact that you decided to expose all exports in this file created new problems

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**